xbmc/addons/AddonDll.cpp

```diff
@@ -425,5 +523,233 @@ bool CAddonDll::CheckAPIVersion(int type)
+  m_interface.firstKodiInstance = firstKodiInstance;
+
+  m_interface.toKodi.kodiBase = this;
+  m_interface.toKodi.get_addon_path = get_addon_path;
+  m_interface.toKodi.get_base_user_path = get_base_user_path;
+  m_interface.toKodi.addon_log_msg = addon_log_msg;
+  m_interface.toKodi.get_setting = get_setting;
+  m_interface.toKodi.set_setting = set_setting;
+  m_interface.toKodi.free_string = free_string;
+
+  return true;
+}
+
+void CAddonDll::DeInitInterface()
+{
+  if (m_interface.libBasePath)
+    free((char*)m_interface.libBasePath);
+  m_interface.libBasePath = nullptr;
+}
+  
+char* CAddonDll::get_addon_path(void* kodiBase)
+{ 
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr)
+  {
+    CLog::Log(LOGERROR, "get_addon_path(...) called with empty kodi instance pointer");
+    return nullptr;
+  }
+
+  return strdup(CSpecialProtocol::TranslatePath(addon->Path()).c_str());
+}
+
+char* CAddonDll::get_base_user_path(void* kodiBase)
+{ 
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr)
+  {
+    CLog::Log(LOGERROR, "get_base_user_path(...) called with empty kodi instance pointer");
+    return nullptr;
+  }
+
+  return strdup(CSpecialProtocol::TranslatePath(addon->Profile()).c_str());
+}
+
+void CAddonDll::addon_log_msg(void* kodiBase, const int addonLogLevel, const char* strMessage)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr)
+  {
+    CLog::Log(LOGERROR, "addon_log_msg(...) called with empty kodi instance pointer");
+    return;
+  }
+
+  int logLevel = LOGNONE;
+  switch (addonLogLevel)
+  {
+    case ADDON_LOG_FATAL:
+      logLevel = LOGFATAL;
+      break;
+    case ADDON_LOG_SEVERE:
+      logLevel = LOGSEVERE;
+      break;
+    case ADDON_LOG_ERROR:
+      logLevel = LOGERROR;
+      break;
+    case ADDON_LOG_WARNING:
+      logLevel = LOGWARNING;
+      break;
+    case ADDON_LOG_NOTICE:
+      logLevel = LOGNOTICE;
+      break;
+    case ADDON_LOG_INFO:
+      logLevel = LOGINFO;
+      break;
+    case ADDON_LOG_DEBUG:
+      logLevel = LOGDEBUG;
+      break;
+    default:
+      break;
+  }
+
+  CLog::Log(logLevel, "AddOnLog: %s: %s", addon->Name().c_str(), strMessage);
+}
+
+bool CAddonDll::get_setting(void* kodiBase, const char* settingName, void* settingValue)
+{ 
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (addon == nullptr || settingName == nullptr || settingValue == nullptr)
+  {
+    CLog::Log(LOGERROR, "kodi::General::%s - invalid data (addon='%p', settingName='%p', settingValue='%p')",
+                                        __FUNCTION__, addon, settingName, settingValue);
+
+    return false;
+  }
+
+  CLog::Log(LOGDEBUG, "kodi::General::%s - add-on '%s' requests setting '%s'",
+                        __FUNCTION__, addon->Name().c_str(), settingName);
+
+  if (strcasecmp(settingName, "__addonpath__") == 0)
+  {
+    strcpy((char*) settingValue, addon->Path().c_str());
```

Comment:
don't use strcpy, so easy to mess up and write random memory.
get_Setting need to allocate the string here and return it or know about the size but preferably allocate it.